### PR TITLE
[oneDNN] Concat refactoring and disabling caching 

### PR DESCRIPTION
**Pull request by jczaja**
### PR types
Others

### PR changes
OPs

### Describe
This PR disabled oneDNN primitives caching by PaddlePaddle in favour of oneDNN its own caching mechanism. Also refactoring to Acquire API was made for concat

### CODE REVIEW

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 1]
```diff
@@ -29,6 +29,72 @@ using mkldnn::concat;
 using mkldnn::stream;
 using platform::to_void_cast;
 
+template <typename T>
+class ConcatMKLDNNHandler
+    : public platform::MKLDNNHandlerNoCachingT<T, dnnl::concat> {
+ public:
+  ConcatMKLDNNHandler(const paddle::framework::ExecutionContext& ctx,
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
  ConcatMKLDNNHandler(const framework::ExecutionContext& ctx,
```

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 2]
```diff
@@ -29,6 +29,72 @@ using mkldnn::concat;
...
+      }
+      output->Resize(out_dims);
+    }
+
+    if (concat_axis < 0) {
+      concat_axis = concat_axis + rank;
+    }
+
+    memory::data_type dt =
+        paddle::framework::ToMKLDNNDataType(inputs[0]->type());
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
        framework::ToMKLDNNDataType(inputs[0]->type());
```

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 3]
```diff
@@ -29,6 +29,72 @@ using mkldnn::concat;
...
+    std::vector<memory::desc> srcs_md;
+    srcs_md.reserve(inputs.size());
+
+    // Create memory descriptors for each of inputs
+    for (size_t i = 0; i < inputs.size(); i++) {
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
    for (size_t i = 0; i < inputs.size(); ++i) {
```

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 4]
```diff
@@ -29,6 +29,72 @@ using mkldnn::concat;
...
+      const auto dims =
+          paddle::framework::vectorize<int64_t>(inputs[i]->dims());
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
          framework::vectorize<int64_t>(inputs[i]->dims());
```

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 5]
```diff
@@ -29,6 +29,72 @@ using mkldnn::concat;
...
+      srcs_md.emplace_back(memory::desc(dims, dt, inputs[i]->format()));
+    }
+
+    auto dst_dims = paddle::framework::vectorize<int64_t>(output->dims());
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
    auto dst_dims = framework::vectorize<int64_t>(output->dims());
```

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 6]
```diff
@@ -29,6 +29,72 @@ using mkldnn::concat;
...
+    auto dst_md = memory::desc(dst_dims, dt, MKLDNNMemoryFormat::any);
+
+    this->AcquireForwardPrimitiveDescriptor(dst_md, concat_axis, srcs_md);
+  }
+
+  // (jczaja) concat oneDNN prim is not having .desc attribute so
+  // we cannot use base AcquireForwardPrimitiveDescriptor
+  void AcquireForwardPrimitiveDescriptor(
+      const mkldnn::memory::desc& dst_md, const int concat_axis,
+      const std::vector<mkldnn::memory::desc>& srcs_md) {
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
      const std::vector<memory::desc>& srcs_md) {
```

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 7]
```diff
@@ -29,6 +29,72 @@ using mkldnn::concat;
...
+    auto dst_dims = paddle::framework::vectorize<int64_t>(output->dims());
+    auto dst_md = memory::desc(dst_dims, dt, MKLDNNMemoryFormat::any);
+
+    this->AcquireForwardPrimitiveDescriptor(dst_md, concat_axis, srcs_md);
+  }
+
+  // (jczaja) concat oneDNN prim is not having .desc attribute so
+  // we cannot use base AcquireForwardPrimitiveDescriptor
+  void AcquireForwardPrimitiveDescriptor(
+      const mkldnn::memory::desc& dst_md, const int concat_axis,
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
      const memory::desc& dst_md, const int concat_axis,
```

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 8]
```diff
@@ -29,6 +29,72 @@ using mkldnn::concat;
...
+      const std::vector<mkldnn::memory::desc>& srcs_md) {
+    this->fwd_pd_.reset(new dnnl::concat::primitive_desc(
+        dst_md, concat_axis, srcs_md, this->engine_));
+  }
+
+  std::shared_ptr<mkldnn::memory> AcquireSrcMemory(
+      const framework::Tensor& input, int i) {
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
      const Tensor& input, int i) {
```

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc [hunk 9]
```diff
@@ -72,160 +116,32 @@ static const std::vector<const Tensor*> ReduceMultiInput(
...
-  std::vector<mkldnn::memory> srcs;
-  paddle::optional<mkldnn::memory> dst_mem;
-};
-
 template <typename T>
 class ConcatMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
  public:
   void Compute(const paddle::framework::ExecutionContext& ctx) const override {
+    auto& dev_ctx =
+        ctx.template device_context<paddle::platform::MKLDNNDeviceContext>();
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]:
```suggestion
        ctx.template device_context<platform::MKLDNNDeviceContext>();
```

**Review by jakpiase — COMMENTED** [review]
I really like the initiative to change from factory to acquireAPI :)

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]: ok

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]: ok

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]: ok

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc (hunk 7 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]: ok

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc (hunk 6 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]: ok

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc (hunk 8 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]: ok

**Comment by jczaja** [comment]
@jakpiase Please continue your review

File: paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc (hunk 9 above)
Comment (jakpiase) [reply paddle/fluid/operators/mkldnn/concat_mkldnn_op.cc]: You have forgotten about this one

**Review by jakpiase — DISMISSED** [review]
LGTM, I really appreciate your effort to unify our operators

**Review by lidanqing-vv — DISMISSED** [review]
LGTM

**Review by lidanqing-vv — APPROVED** [review]
LGTM

**Pull request merged**